awss3/store.go

```diff
@@ -0,0 +1,703 @@
+	default:
+		return nil, nil, ErrNoAuth
+	}
+
+	if conf.BaseUrl != "" {
+		awsConf.WithEndpoint(conf.BaseUrl).WithS3ForcePathStyle(true)
+	}
+
+	disableSSL := conf.Settings.Bool(ConfKeyDisableSSL)
+	if disableSSL {
+		awsConf.WithDisableSSL(true)
+	}
+
+	sess := session.New(awsConf)
+	if sess == nil {
+		return nil, nil, ErrNoS3Session
+	}
+
+	s3Client := s3.New(sess)
+
+	return s3Client, sess, nil
+}
+
+// NewStore Create AWS S3 storage client of type cloudstorage.Store
+func NewStore(c *s3.S3, sess *session.Session, conf *cloudstorage.Config) (*FS, error) {
+
+	if conf.TmpDir == "" {
+		return nil, fmt.Errorf("unable to create cachepath. config.tmpdir=%q", conf.TmpDir)
+	}
+	err := os.MkdirAll(path.Dir(conf.TmpDir), 0775)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create cachepath. config.tmpdir=%q err=%v", conf.TmpDir, err)
+	}
+
+	uid := uuid.NewUUID().String()
+	uid = strings.Replace(uid, "-", "", -1)
+
+	return &FS{
+		client:    c,
+		sess:      sess,
+		bucket:    conf.Bucket,
+		cachepath: conf.TmpDir,
+		ID:        uid,
+		PageSize:  cloudstorage.MaxResults,
+	}, nil
+}
+
+// Type of store = "s3"
+func (f *FS) Type() string {
+	return StoreType
+}
+
+// Client gets access to the underlying google cloud storage client.
+func (f *FS) Client() interface{} {
+	return f.client
+}
+
+// String function to provide s3://..../file   path
+func (f *FS) String() string {
+	return fmt.Sprintf("s3://%s/", f.bucket)
+}
+
+// NewObject of Type s3.
+func (f *FS) NewObject(objectname string) (cloudstorage.Object, error) {
+	obj, err := f.Get(context.Background(), objectname)
+	if err != nil && err != cloudstorage.ErrObjectNotFound {
+		return nil, err
+	} else if obj != nil {
+		return nil, cloudstorage.ErrObjectExists
+	}
+
+	cf := cloudstorage.CachePathObj(f.cachepath, objectname, f.ID)
+
+	return &object{
+		fs:         f,
+		name:       objectname,
+		metadata:   map[string]string{cloudstorage.ContentTypeKey: cloudstorage.ContentType(objectname)},
+		bucket:     f.bucket,
+		cachedcopy: nil,
+		cachepath:  cf,
+	}, nil
+}
+
+// Get a single File Object
+func (f *FS) Get(ctx context.Context, objectpath string) (cloudstorage.Object, error) {
+
+	obj, err := f.getObject(ctx, objectpath)
+	if err != nil {
+		return nil, err
+	} else if obj == nil {
+		return nil, cloudstorage.ErrObjectNotFound
+	}
+
+	return obj, nil
+}
+
+// get single object
+func (f *FS) getObject(ctx context.Context, objectname string) (*object, error) {
+
+	res, err := f.client.GetObjectWithContext(ctx, &s3.GetObjectInput{
+		Key:    aws.String(objectname),
+		Bucket: aws.String(f.bucket),
+	})
+	if err != nil {
+		// translate the string error to typed error
+		if strings.Contains(err.Error(), "NoSuchKey") {
+			return nil, cloudstorage.ErrObjectNotFound
+		}
+		return nil, err
+	}
+	// TODO:  this is messed up, we need to leave it open for some?
+	// should we not even have a helper function?
+	res.Body.Close()
```

Comment:
interesting...  This is a case were writing GCS first influenced our API.  In GCS you first get the Object which is just metadata, then you use that object to create a reader over it.   It appears that in S3 they do both in a single operation.

Reply:
I agree, seems super bad api to me.  should i convert over to using the List api maybe?

Reply:
found a different api HEAD to use